### PR TITLE
Improve DataTable types

### PR DESCRIPTION
**Pull request by avin-kavish**
**What does this PR do?** 

Improve types of DataTable

- Added better types to event handlers
- Remove duplicated types by re-using types from `utils`
- Group props by type
- Added generics for row data type
- Pretty print type definitions with prettier

**Where should the reviewer start?** `DataTable/index.d.ts` 

**What testing has been done on this PR?** Used in a linked project.

**How should this be manually tested?** Use it I guess. Maybe we should write some unit tests in TypeScript.

**What are the relevant issues?** #3951

**Do the grommet docs need to be updated?** No

**Should this PR be mentioned in the release notes?** If you want :)

#### Is this change backwards compatible or is it a breaking change?

This might cause some type errors in loosely written TypeScript code.

### CODE REVIEW

File: src/js/components/DataTable/index.d.ts [hunk 1]
```diff
@@ -1,53 +1,92 @@
...
-} from "../../utils";
+  BackgroundType,
+  PadType,
+  BorderType,
+} from '../../utils';
 
-export interface DataTableProps {
+type Sections<T> = { header?: T; body?: T; footer?: T };
+
+type ColumnSizeType =
```
Comment (ShimiSun) [inline src/js/components/DataTable/index.d.ts:14]: Should this be exported so users can use access it directly?

File: src/js/components/DataTable/index.d.ts [hunk 2]
```diff
@@ -1,53 +1,92 @@
...
+    sortable?: boolean;
+    size?: ColumnSizeType | string;
+    verticalAlign?: 'middle' | 'top' | 'bottom';
+  }[];
+  pad?: PadType | Sections<PadType>;
   resizeable?: boolean;
-  rowProps?: { [_:string]: { background?: {}, border?: {}, pad?: {}}};
-  size?: "small" | "medium" | "large" | "xlarge" | string;
-  sort?: { property: string, direction: "asc" | "desc" };
+  size?: 'small' | 'medium' | 'large' | 'xlarge' | string;
```
Comment (ShimiSun) [inline src/js/components/DataTable/index.d.ts:57]: I like the new structure you are offering, and I think the order within the structure should be alphabetical for consistency.
this line could be moved to line 66.

**Comment by ShimiSun** [comment]
Also, there is a breaking change on this PR, take a look at the current DataTable tsx file that got broken 
https://app.circleci.com/pipelines/github/grommet/grommet/4124/workflows/62bfb77e-ff92-4dc9-8167-fa2ba9aff6e7/jobs/18829

**Comment by avin-kavish** [comment]
Hmm.. I can't access that link, but I'll take a look

**Comment by avin-kavish** [comment]
Hmm... shouldn't be, looks like common types are slightly different to the local ones. I'll fix that.

File: src/js/components/DataTable/index.d.ts (hunk 1 above)
Comment (avin-kavish) [reply src/js/components/DataTable/index.d.ts]: Yeah, few of it can be

**Comment by avin-kavish** [comment]
> I would like to test these changes with a 'tsx' file of storybook to make sure there are no regressions.

I was thinking of making a copy of `data.js` and adding type annotations to it.

**Comment by avin-kavish** [comment]
I'm getting this error when I try to add a new file to Storybook. Is there something special I need to do?

```
ERROR in ./src/js/components/DataTable/stories/typescript/Clickable.tsx
Module not found: Error: Can't resolve './data' in '.../grommet-original/src/js/components/DataTable/stories/typescript'
 @ ./src/js/components/DataTable/stories/typescript/Clickable.tsx 9:0-39 12:50-57 12:65-69
 @ ./src/js sync \.stories\.js$|\/stories\/.*\.js$|\/stories\/.*\.ts$|\/stories\/.*\.tsx$
 @ ./storybook/config.js
 @ multi ./node_modules/@storybook/core/dist/server/common/polyfills.js ./node_modules/@storybook/core/dist/server/preview/globals.js ./storybook/config.js (webpack)-hot-middleware/client.js?reload=true
```

**Comment by ShimiSun** [comment]
Where are you trying to add the file? it should be under DataTable/stories/typescript/avin.tsx
>I was thinking of making a copy of data.js and adding types to it, combined with the current tsx files, that should be enough.

That would be great, thanks! I'm thinking you can leverage the existing typescript story instead of adding a new one, and add props that you edited on the interface there for the purpose of type checking.

@avin-kavish if the `data` file is located on the stories directory and not on the typescript directory, then you'll need to point .'./data'.

**Comment by avin-kavish** [comment]
lol. Did you just use my name as the filename? :laughing:

Hmm... I don't know, the problem has not gone away, even though all the imports are correct according to VSCode.

**Comment by ShimiSun** [comment]
I've cleaned up this PR and I think it is ready to go.
Tested manually there are no typescript failures with each DataTable story.

**Comment by ShimiSun** [comment]
Thank you for your contribution!

**Pull request merged**